atef/widgets/config/find_replace.py

```diff
@@ -0,0 +1,1009 @@
+                parent=self,
+                caption='Select a config',
+                filter='Json Files (*.json)',
+            )
+        if not filename:
+            return
+
+        def finish_setup():
+            self.details_list.clear()
+            self.setup_edits_table()
+            self.setup_devices_list()
+            self.update_title()
+
+        self.busy_thread = BusyCursorThread(
+            func=partial(self.load_file, filepath=filename)
+        )
+        self.busy_thread.task_finished.connect(finish_setup)
+        self.busy_thread.start()
+
+    def load_file(self, filepath: str) -> Union[ConfigurationFile, ProcedureFile]:
+        with open(filepath, 'r') as fp:
+            self._original_json = json.load(fp)
+
+        try:
+            data = deserialize(ConfigurationFile, self._original_json)
+        except ValidationError:
+            logger.debug('failed to open as passive checkout')
+            try:
+                data = deserialize(ProcedureFile, self._original_json)
+            except ValidationError:
+                logger.error('failed to open file as either active '
+                             'or passive checkout')
+
+        self.fp = filepath
+        self.orig_file = data
+
+    def setup_devices_list(self) -> None:
+        """
+        Set up the devices list.  Runs the preparation and cache inspection
+        in a BusyCursorThread, and fills the devices list after its completion
+        """
+        self.busy_thread = BusyCursorThread(func=self._get_devices_in_file)
+        self.busy_thread.task_finished.connect(self._fill_devices_list)
+        self.busy_thread.start()
+
+    def _get_devices_in_file(self) -> None:
+        """
+        Gather devices and signal in the original file by preparing it and
+        inspecting caches.  Temporarily clears the happi loader cache, but
+        actually clears the DataCache
+        """
+        with patch_client_cache():
+            client = get_happi_client()
+            if isinstance(self.orig_file, ConfigurationFile):
+                prep_file = PreparedFile.from_config(self.orig_file, client=client)
+                asyncio.run(prep_file.fill_cache())
+            elif isinstance(self.orig_file, ProcedureFile):
+                prep_file = PreparedProcedureFile.from_origin(self.orig_file)
+
+            cache = get_signal_cache()
+            self._signals = list(cache.keys())
+            self._devices = list(happi.loader.cache.keys())
+
+    def _fill_devices_list(self) -> None:
+        self.device_table.setRowCount(max(len(self._signals), len(self._devices)))
+        for i, sig in enumerate(self._signals):
+            self.device_table.setItem(i, 1, QtWidgets.QTableWidgetItem(sig))
+        for i, dev in enumerate(self._devices):
+            self.device_table.setItem(i, 0, QtWidgets.QTableWidgetItem(dev))
+
+    def verify_changes(self) -> None:
+        verify_changes(self, self.orig_file)
+
+    def save_file(self) -> None:
```

Comment:
Shouldn't there be an implicit "verify_changes" call before it's saved?

Reply:
I agree with verifying as long as it doesn't prevent saving, since users might use this in an environment where they don't have access to all the necessary devices, making preparation fail.

Reply:
Something like "Verification failed! Save anyway? [y/n]"